### DISCUSSION TITLE
Compiling for Linux

### DISCUSSION THREAD

**Discussion by EnfrigFalco** [Markemp/Cryengine-Converter#164] [Q&A]
Need help with compiling for Linux. I have successfully compiled StarFab but it is useless without Cryengine-Converter. Can anyone tell me what I am doing wrong? 

I compiled Cryengine-Converter on Linux (Fedora 38) with following commands:
```
# Install dotnet
sudo dnf -y install dotnet-sdk-6.0 dotnet-sdk-7.0

# Clone
git clone https://github.com/Markemp/Cryengine-Converter.git
cd Cryengine-Converter/cgf-converter/

# Compile
dotnet publish cgf-converter.csproj -f net6.0 -r linux-x64 --self-contained true
```
However the bin and dll will not detect any input file
```
./cgf-converter AEGS_Avenger.cga 
Using up to 12 threads
No corresponding input file exist: AEGS_Avenger.cga
[INF:Program] Finished.
```
Same with dll
```
dotnet cgf-converter.dll AEGS_Avenger.cga 
Using up to 12 threads
No corresponding input file exist: AEGS_Avenger.cga
[INF:Program] Finished.
```

I compiled and tested from Linux to Windows with `-r win-x64`  successfully, so I believe it is compiling correctly. 

Is there some trick with the dotnet command-line on Linux? what am I missing? has anybody been successful in running converter natively on Linux?

**Comment by Markemp** [comment]
There are 2 paths you need to have set up to get this to work.  The first is the one to the `cgf-converter.exe` program.  It looks like your terminal is able to find the path to that though, since it's running instead of giving you a file not found error.

The other is the path to the game files.  For the most part, the converter program expects the full game files to be extracted into a directory structure that matches the game hierarchy.  In my setup, I extract everything to a `d:\depot\sc` directory, so under there you will see an `objects`, `textures`, and a few other directories.  All the game files are under that.  That is called the object directory, or data directory.  Data directory is probably a better name. 

In your case, it can't find the .cga file.  So you either need to reference the .cga file by the full path, or add the path to the cgf-converter program into your terminal Path, and then you can navigate to any subdirectory of the data directory and just run it that way.

Let me know if this helps, or if you need more info.  I haven't heard of anyone using linux to do this, so I'm excited to see how it turns out.  Part of the benefit of this program being written in .Net Core is that this is theoretically possible, but I haven't heard of anyone actually doing yet, and would love to see how it works out.

**Reply by OmegaRogue** [reply]
Even with a full path, it doesn't work

**Reply by OmegaRogue** [reply]
I'm also getting the same issue on windows with the 1.6.0 preview build

**Reply by Markemp** [reply]
Weird that it isn't working on windows either.  I just tested running the script from its directory and pointing to the file over in the depot, and it was able to find it.

```
 Geoff    scripts  ﮫ5ms⠀   .\cgf-converter.exe D:\depot\mwo\objects\purchasable\cockpit_hanging\50calnecklace\50calnecklace_a.cga
Using up to 16 threads
Found input: D:\depot\mwo\objects\purchasable\cockpit_hanging\50calnecklace\50calnecklace_a.cga
[INF:Program] [1/1 0.00%] D:\depot\mwo\objects\purchasable\cockpit_hanging\50calnecklace\50calnecklace_a.cga
[INF:50calnecklace_a.cga] Found following potential material files.  If you are not specifying a material file and the materials don't look right, trying one of the following files:
[INF:50calnecklace_a.cga]    Objects/purchasable/cockpit_hanging/50calnecklace/50calnecklace_a
[INF:Program] Finished.
```
Can you provide the full command line that you're using to run the program?

**Reply by ZoldorfTheWizard** [reply]
Heyo!

I'm playing around with getting Starfab exporting to work on Linux.  This project is the last one to cross that line so was kinda ecstatic to see it was built on dotnet. I am quite familiar with dotnet/C# so I pulled down this repo, saw the dependencies were mostly cross platform at a glance and built it for Linux with:  

```
dotnet build 
dotnet publish --runtime linux-x64 
cp cgf-converter/bin/Release/net8.0/linux-x64/publish/* /usr/bin
```

```
/apollo/Starfab Exports/
❯ cgf-converter                                                                                                                                      

cgf-converter [-usage] | <.cgf file> [-outputfile <output file>] [-dae] [-obj] [-glb] [-gltf] [-notex/-png/-tif/-tga] [-group] [-excludenode <nodename>] [-excludemat <matname>] [-loglevel <LogLevel>] [-throw] [-dump] [-objectdir <ObjectDir>]

CryEngine Converter v1.6.0.0

-usage:            Prints out the usage statem...
```
However it seems like I am running into the same issue as the original poster where the converter cannot find the input file/s.  

I have the directory setup like the converter expects and have provided full paths like you suggested but it seems to not be finding the file either

```
/apollo/Starfab Exports/DRAK_Cutlass_Red/Data 
❯ ls  -la
total 36
drwxr-xr-x  9 zoldorf zoldorf 4096 Jul 31 21:20 .
drwxr-xr-x  3 zoldorf zoldorf 4096 Jul 31 21:20 ..
drwxr-xr-x  3 zoldorf zoldorf 4096 Jul 31 21:20 Animations
drwxr-xr-x  3 zoldorf zoldorf 4096 Jul 31 21:20 libs
drwxr-xr-x  4 zoldorf zoldorf 4096 Jul 31 21:20 Materials
drwxr-xr-x  3 zoldorf zoldorf 4096 Jul 31 21:20 ObjectContainers
drwxr-xr-x  4 zoldorf zoldorf 4096 Jul 31 21:20 Objects
drwxr-xr-x  3 zoldorf zoldorf 4096 Jul 31 21:20 Scripts
drwxr-xr-x 11 zoldorf zoldorf 4096 Jul 31 21:20 Textures
```
```
/apollo/Starfab Exports/DRAK_Cutlass_Red/Data 
❯ cgf-converter /apollo/Starfab\ Exports/DRAK_Cutlass_Red/Data/Objects/Spaceships/Ships/DRAK/Cutlass/Cutlass_Red/DRAK_Cutlass_Red_Airlock_Sleeve.cga 
Using up to 32 threads
No corresponding input file exist: /apollo/Starfab Exports/DRAK_Cutlass_Red/Data/Objects/Spaceships/Ships/DRAK/Cutlass/Cutlass_Red/DRAK_Cutlass_Red_Airlock_Sleeve.cga
[INF:Program] Finished. Rendered 0 file(s)
```

I have also put this directory (/apollo/Starfab Exports/DRAK_Cutlass_Red/Data) into my path but that also seems to not work. 

Any Ideas? Ill poke deeper for now and try to debug/ see if I can maybe see whats causing the file to not be found.

Thanks for the awesome tool!
